crates/bevy_ecs/src/result/assume.rs

```diff
@@ -0,0 +1,59 @@
+use super::Error;
+
+/// Assume that `Self<T>` is `T`, otherwise return the provided error.
+pub trait Assume<T> {
+    /// The error type returned by [`Assume::assume`].
```

Comment:
While I don't think we should make it a trait bound, we should mention that these types should typically implement the `Error` trait.